File: packages/kits/PersistenceKit/Sources/PersistenceKit/Predicate.swift
// StoragePredicate.swift
//
// StoragePredicate tree per DECISION_STORAGEKIT_DESIGN §4 (Q2).
// Closed enum, three operator families: logical, comparison, bitmap.
// BitmapEvaluator compiles Filter → StoragePredicate; backend compiles
// StoragePredicate → backend-native SQL. PersistenceKit treats predicates
// as opaque except for compilation.

import Foundation

public indirect enum StoragePredicate: Sendable {
    // Logical
    case and([StoragePredicate])
    case or([StoragePredicate])
    case not(StoragePredicate)
    case isTrue
    case isFalse

    // Comparison
    case eq(Column, TypedValue)
    case neq(Column, TypedValue)
    case lt(Column, TypedValue)
    case lte(Column, TypedValue)
    case gt(Column, TypedValue)
    case gte(Column, TypedValue)
    case isNull(Column)
    case isNotNull(Column)
    case `in`(Column, [TypedValue])
    case like(Column, String)

    // Bitmap (Int64 columns only)
    case bitmaskAll(Column, mask: Int64)
    case bitmaskAny(Column, mask: Int64)
    case bitmaskNone(Column, mask: Int64)
    case bitwiseEq(Column, expected: Int64, mask: Int64)
}

public extension StoragePredicate {
    /// Convenience: combine predicates with AND, short-circuiting
    /// trivial cases.
    static func all(_ predicates: [StoragePredicate]) -> StoragePredicate {
        let filtered = predicates.filter {
            if case .isTrue = $0 { return false }
            return true
        }
        if filtered.isEmpty { return .isTrue }
        if filtered.count == 1 { return filtered[0] }
        if filtered.contains(where: { if case .isFalse = $0 { return true }; return false }) {
            return .isFalse
        }
        return .and(filtered)
    }

    /// Convenience: combine predicates with OR.
    static func any(_ predicates: [StoragePredicate]) -> StoragePredicate {
        let filtered = predicates.filter {
            if case .isFalse = $0 { return false }
            return true
        }
        if filtered.isEmpty { return .isFalse }
        if filtered.count == 1 { return filtered[0] }
        if filtered.contains(where: { if case .isTrue = $0 { return true }; return false }) {
            return .isTrue
        }
        return .or(filtered)
    }
}

public enum OrderDirection: Sendable {
    case ascending
    case descending
}

public struct OrderClause: Sendable {
    public let column: Column
    public let direction: OrderDirection

    public init(column: Column, direction: OrderDirection = .ascending) {
        self.column = column
        self.direction = direction
    }
}

File: packages/kits/PersistenceKit/rust/src/predicate.rs
//! Closed-enum predicate algebra.
//!
//! Mirror of Swift's `StoragePredicate`. Three operator families:
//! logical, comparison, bitmap. Backends compile to native
//! query languages; the kit treats the predicate as opaque
//! except for compilation.

use crate::types::{Column, TypedValue};

#[derive(Debug, Clone)]
pub enum StoragePredicate {
    // Logical
    And(Vec<StoragePredicate>),
    Or(Vec<StoragePredicate>),
    Not(Box<StoragePredicate>),
    IsTrue,
    IsFalse,

    // Comparison
    Eq(Column, TypedValue),
    Neq(Column, TypedValue),
    Lt(Column, TypedValue),
    Lte(Column, TypedValue),
    Gt(Column, TypedValue),
    Gte(Column, TypedValue),
    IsNull(Column),
    IsNotNull(Column),
    In(Column, Vec<TypedValue>),
    Like(Column, String),

    // Bitmap (Int64 / Bitmap columns only)
    BitmaskAll {
        column: Column,
        mask: i64,
    },
    BitmaskAny {
        column: Column,
        mask: i64,
    },
    BitmaskNone {
        column: Column,
        mask: i64,
    },
    BitwiseEq {
        column: Column,
        expected: i64,
        mask: i64,
    },
}

impl StoragePredicate {
    /// Combine predicates with AND, short-circuiting trivial
    /// cases. Mirror of Swift's `StoragePredicate.all`.
    pub fn all(predicates: Vec<StoragePredicate>) -> StoragePredicate {
        let mut filtered: Vec<StoragePredicate> = predicates
            .into_iter()
            .filter(|p| !matches!(p, StoragePredicate::IsTrue))
            .collect();
        if filtered.is_empty() {
            return StoragePredicate::IsTrue;
        }
        if filtered
            .iter()
            .any(|p| matches!(p, StoragePredicate::IsFalse))
        {
            return StoragePredicate::IsFalse;
        }
        if filtered.len() == 1 {
            return filtered.remove(0);
        }
        StoragePredicate::And(filtered)
    }

    /// Combine predicates with OR. Mirror of Swift's
    /// `StoragePredicate.any`.
    pub fn any(predicates: Vec<StoragePredicate>) -> StoragePredicate {
        let mut filtered: Vec<StoragePredicate> = predicates
            .into_iter()
            .filter(|p| !matches!(p, StoragePredicate::IsFalse))
            .collect();
        if filtered.is_empty() {
            return StoragePredicate::IsFalse;
        }
        if filtered
            .iter()
            .any(|p| matches!(p, StoragePredicate::IsTrue))
        {
            return StoragePredicate::IsTrue;
        }
        if filtered.len() == 1 {
            return filtered.remove(0);
        }
        StoragePredicate::Or(filtered)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OrderDirection {
    Ascending,
    Descending,
}

#[derive(Debug, Clone)]
pub struct OrderClause {
    pub column: Column,
    pub direction: OrderDirection,
}

impl OrderClause {
    pub fn new(column: Column, direction: OrderDirection) -> Self {
        OrderClause { column, direction }
    }

    pub fn ascending(column: Column) -> Self {
        OrderClause {
            column,
            direction: OrderDirection::Ascending,
        }
    }

    pub fn descending(column: Column) -> Self {
        OrderClause {
            column,
            direction: OrderDirection::Descending,
        }
    }
}
